spikeinterface/sortingcomponents/waveforms/neural_network_denoiser.py

```diff
@@ -0,0 +1,118 @@
+        with open(json_file_path, "r") as json_file:
+            peak_interval_dict = json.load(json_file)
+
+        model_ms_before = peak_interval_dict["ms_before"]
+        model_ms_after = peak_interval_dict["ms_after"]
+        model_sampling_frequency = peak_interval_dict["sampling_frequency"]
+
+        ms_before_mismatch = waveforms_ms_before != model_ms_before
+        ms_after_missmatch = waveforms_ms_after != model_ms_after
+        sampling_frequency_mismatch = waveforms_sampling_frequency != model_sampling_frequency
+        if ms_before_mismatch or ms_after_missmatch or sampling_frequency_mismatch:
+            exception_string = (
+                "Model and waveforms mismatch \n"
+                f"{model_ms_before=} and {waveforms_ms_after=} \n"
+                f"{model_ms_after=} and {waveforms_ms_after=} \n"
+                f"{model_sampling_frequency=} and {waveforms_sampling_frequency=} \n"
+            )
+            raise ValueError(exception_string)
+
+    def load_model(self):
+        repo_id = "SpikeInterface/test_repo"
+        subfolder = "mearec_toy_model"
+        filename = "toy_model_marec.pt"
+
+        model_path = hf_hub_download(repo_id=repo_id, subfolder=subfolder, filename=filename)
+        denoiser = SingleChannel1dCNNDenoiser(pretrained_path=model_path, spike_size=128)
+        denoiser = denoiser.load()
+
+        return denoiser
+
+    def compute(self, traces, peaks, waveforms):
+        num_channels = waveforms.shape[2]
+
+        # Collapse channels and transform to torch tensor
+        temporal_waveforms = to_temporal_representation(waveforms)
+        temporal_waveforms_tensor = torch.from_numpy(temporal_waveforms).float()
+
+        # Denoise
+        denoised_temporal_waveforms = self.denoiser(temporal_waveforms_tensor).detach().numpy()
+
+        # Reconstruct representation with channels
+        desnoised_waveforms = from_temporal_representation(denoised_temporal_waveforms, num_channels)
+
+        return desnoised_waveforms
+
+
+class SingleChannel1dCNNDenoiser(nn.Module):
+    def __init__(self, pretrained_path=None, n_filters=[16, 8], filter_sizes=[5, 11], spike_size=121):
+        super().__init__()
+
+        out_channels_conv1, out_channels_conv_2 = n_filters
+        kernel_size_conv1, kernel_size_conv2 = filter_sizes
+        self.conv1 = nn.Sequential(nn.Conv1d(1, out_channels_conv1, kernel_size_conv1), nn.ReLU())
+        self.conv2 = nn.Sequential(nn.Conv1d(out_channels_conv1, out_channels_conv_2, kernel_size_conv2), nn.ReLU())
+        n_input_feat = out_channels_conv_2 * (spike_size - kernel_size_conv1 - kernel_size_conv2 + 2)
+        self.out = nn.Linear(n_input_feat, spike_size)
+        self.pretrained_path = pretrained_path
+
+    def forward(self, x):
+        x = x[:, None]
+        x = self.conv1(x)
+        x = self.conv2(x)
+        x = x.view(x.shape[0], -1)
+        x = self.out(x)
+        return x
+
+    def load(self, device="cuda"):
```

Comment:
What happens if cuda is not available? Does it automatically use CPU instead?

Reply:
Apparently not ;)

https://github.com/SpikeInterface/spikeinterface/actions/runs/4353692677/jobs/7608059335

Reply:
All right. I have a gpu card on my local machine so I did not catch this one. 

I propose that we delay this till we have some robust mechanism to test whether gpu capabilities are available:
https://github.com/SpikeInterface/spikeinterface/pull/1398

What do you think?

Reply:
Ok for me ;)

Reply:
Cool. Done, this should be ready to go.

Reply:
@h-mayorquin let's keep it in tests for now, since it's still quite experimental. Is that ok?